tensordict/functional.py

```diff
@@ -87,76 +87,92 @@ def pad_sequence(
         TensorDict(
             fields={
-                a: Tensor(shape=torch.Size([2, 4]), device=cpu, dtype=torch.float32, is_shared=False)},
-            batch_size=torch.Size([]),
+                a: Tensor(shape=torch.Size([2, 4, 8]), device=cpu, dtype=torch.float32, is_shared=False),
+                b: Tensor(shape=torch.Size([2, 5, 8]), device=cpu, dtype=torch.float32, is_shared=False),
+                masks: TensorDict(
+                    fields={
+                        a: Tensor(shape=torch.Size([2, 4]), device=cpu, dtype=torch.bool, is_shared=False),
+                        b: Tensor(shape=torch.Size([2, 6]), device=cpu, dtype=torch.bool, is_shared=False)},
+                    batch_size=torch.Size([2]),
+                    device=None,
+                    is_shared=False)},
+            batch_size=torch.Size([2]),
             device=None,
             is_shared=False)
     """
     if not list_of_tensordicts:
         raise RuntimeError("list_of_tensordicts cannot be empty")
+
     # check that all tensordict match
     if return_mask:
-        list_of_tensordicts = [
-            td.clone(False).set("mask", torch.ones(td.shape, dtype=torch.bool))
-            for td in list_of_tensordicts
-        ]
+        for key in _check_keys(
+            list_of_tensordicts, leaves_only=True, include_nested=True
+        ):
+            list_of_tensordicts = [
+                td.clone(False).set(
+                    tuple(["masks"] + list(key)),
+                    torch.ones(td.get(key).shape[pad_dim], dtype=torch.bool),
+                )
+                for td in list_of_tensordicts
+            ]
+
     keys = _check_keys(list_of_tensordicts, leaves_only=True, include_nested=True)
-    shape = max(len(td) for td in list_of_tensordicts)
-    if shape == 0:
-        shape = [
-            len(list_of_tensordicts),
-        ]
-    elif batch_first:
-        shape = [len(list_of_tensordicts), shape]
-    else:
-        shape = [shape, len(list_of_tensordicts)]
+    shape = [
+        len(list_of_tensordicts),
+    ]
+
     if out is None:
         out = TensorDict(
             {}, batch_size=torch.Size(shape), device=device, _run_checks=False
         )
-        for key in keys:
-            try:
-                out.set(
+        set_func = out.set
+    else:
+        set_func = out.set_
+    for key in keys:
+        try:
+            if key[0] != "masks":
+                set_func(
+                    key,
+                    torch.nn.utils.rnn.pad_sequence(
+                        [
+                            td.get(key).transpose(0, pad_dim)
+                            for td in list_of_tensordicts
+                        ],
+                        batch_first=True,
+                        padding_value=padding_value,
+                    ).transpose(1, pad_dim + 1),
+                )
+            else:
+                set_func(
                     key,
                     torch.nn.utils.rnn.pad_sequence(
```

Comment:
I guess we can't use tensordicts with batch_size because we're relying on pad_sequence.
I thought about using `functorch.dim` here to "hide" the first dimensions and use pad_sequence but I get a `RuntimeError: Batching rule not implemented for aten::pad_sequence. We could not generate a fallback.` (cc @zdevito if you're interested: we'd like to use pad_sequence for batches)

The workaround is this: for all tensordicts, create a version of the tensordict where the batch dimensions are permuted to the back of the dims, then permuted back. We can use tensordict.apply all along:


```python
td_a = TensorDict({
    "a": torch.randn(10, 11, 3, 5),
    "b": torch.randn(10, 11, 3, 5),
}, [10, 11])
td_b = TensorDict({
    "a": torch.randn(10, 11, 4, 5),
    "b": torch.randn(10, 11, 4, 5),
}, [10, 11])

def permute_batch_size(tensor, batch_size=[10, 11, ]):
    L = len(batch_size)
    return tensor.permute(*range(L, tensor.ndim), *range(L))
td_a_p = td_a.apply(permute_batch_size, batch_size=[])
td_b_p = td_b.apply(permute_batch_size, batch_size=[])

def pad_seq(*tensors):
    return torch.nn.utils.rnn.pad_sequence(list(tensors), batch_first=True)
td_pad = td_a_p.apply(pad_seq, td_b_p)
def reset_batch_size(tensor, batch_size=[10, 11, ]):
    L = len(batch_size)
    return tensor.permute(0, *range(tensor.ndim - L, tensor.ndim), *range(1, tensor.ndim - L))
td_pad = td_pad.apply(reset_batch_size)

print(td_pad)
```
prints
```
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([2, 10, 11, 4, 5]), device=cpu, dtype=torch.float32, is_shared=False),
        b: Tensor(shape=torch.Size([2, 10, 11, 4, 5]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)
```

which is pretty cool. You don't really need to reset the batch size since we called empty above so our out-container has already all the metadata set.